xdsl/dialects/bufferization.py

```diff
@@ -33,51 +34,41 @@
+        MemRefType[Attribute] | UnrankedMemrefType[Attribute]
+    ]
+
+    def can_infer(self, constraint_names: set[str]) -> bool:
+        return self.memref_constraint.can_infer(constraint_names)
+
     def infer(self, constraint_context: ConstraintContext) -> Attribute:
-        if self.name in constraint_context.variables:
-            m_type = constraint_context.get_variable(self.name)
-            if isa(m_type, MemRefType[Attribute]):
-                return TensorType(m_type.get_element_type(), m_type.get_shape())
-            if isa(m_type, UnrankedMemrefType[Attribute]):
-                return UnrankedTensorType(m_type.element_type)
-        raise ValueError(f"Unexpected {self.name} - cannot infer attribute")
+        memref_type = self.memref_constraint.infer(constraint_context)
+        if isa(memref_type, MemRefType[Attribute]):
+            return TensorType(memref_type.element_type, memref_type.shape)
+        assert isa(memref_type, UnrankedMemrefType[Attribute])
+        return UnrankedTensorType(memref_type.element_type)
+
+    def get_resolved_variables(self) -> set[str]:
+        return self.memref_constraint.get_resolved_variables()
 
     def verify(self, attr: Attribute, constraint_context: ConstraintContext) -> None:
-        if self.name in constraint_context.variables:
-            seen = constraint_context.get_variable(self.name)
-            if not (
-                isinstance(attr, ContainerType)
-                and isinstance(seen, ContainerType)
-                and attr.get_element_type() == seen.get_element_type()
-            ):
-                raise VerifyException(
-                    f"Unexpected {self.name} - cannot verify element type of attribute {attr}"
-                )
-            if (
-                isinstance(attr, ShapedType) != isinstance(seen, ShapedType)
-                or isinstance(attr, ShapedType)
-                and isinstance(seen, ShapedType)
-                and attr.get_shape() != seen.get_shape()
-            ):
-                raise VerifyException(
-                    f"Unexpected {self.name} - cannot verify shape of attribute {attr}"
-                )
-        elif isinstance(attr, ContainerType):
-            self.constraint.verify(attr, constraint_context)
-            constraint_context.set_variable(self.name, attr)
-        else:
-            raise VerifyException(
-                f"Unexpected {self.name} - attribute must be ContainerType"
-            )
+        if isa(attr, TensorType[Attribute]):
+            memref_type = MemRefType(attr.element_type, attr.shape)
+            return self.memref_constraint.verify(memref_type, constraint_context)
+        if isa(attr, UnrankedTensorType[Attribute]):
+            memref_type = UnrankedMemrefType.from_type(attr.element_type)
+            return self.memref_constraint.verify(memref_type, constraint_context)
+        raise VerifyException(f"Expected TensorType or UnrankedTensorType, got {attr}")
```

Comment:
worth adding a test for this?

Reply:
I think it's probably OK to merge without this but if it's quick and easy it would catch this sort of thing in the future, hopefully

Reply:
Added an array of tests